proposals/nnnn-regex-literals.md

```diff
@@ -0,0 +1,485 @@
+
+The compiler will parse the contents of a regex literal using regex syntax outlined in *[Regex Construction][internal-syntax]*, diagnosing any errors at compile time. The capture types and labels are automatically inferred based on the capture groups present in the regex. Regex literals allows editors and source tools to support features such as syntax coloring inside the literal, highlighting sub-structure of the regex, and conversion of the literal to an equivalent result builder DSL (see *[Regex builder DSL][regex-dsl]*).
+
+A regex literal also allows for seamless composition with the Regex DSL, enabling lightweight intermixing of a regex syntax with other elements of the builder:
+
+```swift
+// A regex for extracting a currency (dollars or pounds) and amount from input 
+// with precisely the form /[$£]\d+\.\d{2}/
+let regex = Regex {
+  Capture { /[$£]/ }
+  TryCapture {
+    /\d+/
+    "."
+    /\d{2}/
+  } transform: {
+    Amount(twoDecimalPlaces: $0)
+  }
+}
+```
+
+This flexibility allows for terse matching syntax to be used when it's suitable, and more explicit syntax where clarity and strong types are required.
+
+Due to the existing use of `/` in comment syntax and operators, there are some syntactic ambiguities to consider. While there are quite a few cases to consider, we do not feel that the impact of any individual case is sufficient to disqualify the syntax. Some of these ambiguities require a couple of source breaking language changes, and as such the `/.../` syntax requires upgrading to a new language mode in order to use.
+
+## Detailed design
+
+### Typed captures
+
+Regex literals have their capture types statically determined by the capture groups present. This follows the same inference behavior as [the DSL][regex-dsl], and is explored in more detail in *[Strongly Typed Captures][strongly-typed-captures]*. We are therefore proposing the following inference behavior for regex literals:
+
+- A `Substring` is always present for the entire match.
+- If any captures are present, a tuple is formed with the `Substring`, with subsequent elements representing the capture types. Captures are ordered according to [their numbering][capture-numbering].
+
+The type of a capture is `Substring` by default, however it gets wrapped in an optional if it is not guaranteed to have a value on a successful match. This occurs when it is nested within a quantification that may be zero, e.g `?`, `*`, and `{0,n}`:
+
+```swift
+let r = /([ab]+)?/
+// regex: Regex<(Substring, Substring?)>
+```
+
+it also occurs when it appears as a child of an alternation:
+
+```swift
+let r = /([ab]+)|\d+/
+// regex: Regex<(Substring, Substring?)>
+```
+
+The optional wrapping will nest arbitrarily if the capture is nested within multiple zero-quantifiers or alternations:
+
+```swift
+let r = /(.)*|\d/
+// regex: Regex<(Substring, Substring??)>
+``` 
+
```

Comment:
I don't think this behavior makes sense and I don't think we should let temporary (I hope) limitations from `RegexBuilder` spill over into permanently locking us into the wrong type system representation.

Reply:
My takeaway from the meeting is that we want to type check this as

```swift
// regex: Regex<(Substring, Substring??)>
```

And we'll be exploring ways to explicitly drop a single layer of optionality for the DSL conversion (turtles all the way down)

Reply:
Updated to say that we will apply at most one layer of optionality. Do we want to talk about the DSL conversion in this proposal?

Reply:
Hmm, hard to say. It's not a language/library thing (other than we happen to be using the same library as the parser to achieve it).

Reply:
Hmm yeah, in that case I think it's probably fine to leave as-is